### PR TITLE
feat: add HuggingFace Hub integration for dataset publishing

### PR DESCRIPTION
**Pull request by nabinchha**
## 📋 Summary

This PR adds comprehensive HuggingFace Hub integration, enabling users to publish DataDesigner datasets directly to the HuggingFace Hub with automated dataset card generation, flexible upload options, and robust error handling.

### ✨ Added
- [`HuggingFaceHubClient`](https://github.com/NVIDIA-NeMo/DataDesigner/blob/nmulepati/feat/7-push-to-hf/packages/data-designer/src/data_designer/integrations/huggingface/client.py) - Complete client for uploading datasets to HuggingFace Hub with support for:
  - Parquet and JSON file formats
  - Automatic repository creation
  - Path mapping configuration for column-specific file handling
  - Comprehensive logging and progress tracking
- [`DataDesignerDatasetCard`](https://github.com/NVIDIA-NeMo/DataDesigner/blob/nmulepati/feat/7-push-to-hf/packages/data-designer/src/data_designer/integrations/huggingface/dataset_card.py) - Rich dataset card generator with:
  - Automatic metadata extraction from DataDesigner configs
  - Model information (aliases, inference parameters, system prompts)
  - Column configuration details (samplers, validators, LLM types)
  - Generation statistics and filtering info
  - Custom markdown template rendering
- `push_to_hub()` method on `DatasetCreationResults` - Simple API for publishing results
- Comprehensive test suite with 569+ test lines covering edge cases and error scenarios

### 🔧 Changed
- Enhanced `DatasetCreationResults` class to support HuggingFace publishing workflow

## 🔍 Attention Areas

> ⚠️ **Reviewers:** Please pay special attention to the following:

- [`client.py`](https://github.com/NVIDIA-NeMo/DataDesigner/blob/nmulepati/feat/7-push-to-hf/packages/data-designer/src/data_designer/integrations/huggingface/client.py) (349 lines) - Core upload logic with path mapping and error handling
- [`dataset_card.py`](https://github.com/NVIDIA-NeMo/DataDesigner/blob/nmulepati/feat/7-push-to-hf/packages/data-designer/src/data_designer/integrations/huggingface/dataset_card.py) (139 lines) - Dataset card template rendering and metadata extraction
- [`test_client.py`](https://github.com/NVIDIA-NeMo/DataDesigner/blob/nmulepati/feat/7-push-to-hf/packages/data-designer/tests/integrations/huggingface/test_client.py) (569 lines) - Extensive test coverage for upload scenarios

---

See this data set card as an example published directly from the create results object: https://huggingface.co/datasets/nabinnvidia/multi-lingual-greetings
```
create_result.push_to_hub(
    repo_id="nabinnvidia/multi-lingual-greetings",
    private=False,
    description="This dataset is a test dataset for multi-lingual greetings."
)
```
Screenshot 2026-02-02 at 11 55 14 AM

Closes https://github.com/NVIDIA-NeMo/DataDesigner/issues/7

Draws inspiration on implementation from conversations in this PR: https://github.com/NVIDIA-NeMo/DataDesigner/pull/127

🤖 *Generated with AI*

### CODE REVIEW

File: packages/data-designer/src/data_designer/integrations/huggingface/client.py [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+                try:
+                    self._api.upload_file(
+                        repo_id=repo_id,
+                        path_or_fileobj=tmp_path,
+                        path_in_repo="metadata.json",
+                        repo_type="dataset",
+                        commit_message="Upload metadata.json",
+                    )
+                finally:
+                    Path(tmp_path).unlink()
```
[the remark this answers is not in the document]
Comment (nabinchha) [reply packages/data-designer/src/data_designer/integrations/huggingface/client.py]: fixed in 243c08739844f871d7cfa9b84a5aa4777367df1d

File: packages/data-designer-engine/src/data_designer/engine/dataset_builders/artifact_storage.py [hunk 2]
```diff
@@ -259,7 +259,7 @@ def write_metadata(self, metadata: dict) -> Path:
         """
         self.mkdir_if_needed(self.base_dataset_path)
         with open(self.metadata_file_path, "w") as file:
-            json.dump(metadata, file, indent=4, sort_keys=True)
+            json.dump(metadata, file, indent=2, sort_keys=True)
```
Comment (johnnygreco) [inline packages/data-designer-engine/src/data_designer/engine/dataset_builders/artifact_storage.py:265]: don't like 4? all good just curious

File: packages/data-designer/src/data_designer/integrations/huggingface/__init__.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+from data_designer.integrations.huggingface.client import HuggingFaceHubClient, HuggingFaceUploadError
+from data_designer.integrations.huggingface.dataset_card import DataDesignerDatasetCard
+
+__all__ = ["HuggingFaceHubClient", "HuggingFaceUploadError", "DataDesignerDatasetCard"]
```
Comment (johnnygreco) [inline packages/data-designer/src/data_designer/integrations/huggingface/__init__.py]: nit: I've started to like the `HuggingFaceHubClient.__name__` pattern. We're not consistent throughout, though

File: packages/data-designer/src/data_designer/integrations/huggingface/client.py [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+        try:
+            self._upload_dataset_card(repo_id, base_dataset_path, description)
+        except Exception as e:
+            raise HuggingFaceUploadError(f"Failed to upload dataset card: {e}") from e
+
+        self._upload_main_dataset_files(repo_id, base_dataset_path)
+        self._upload_processor_files(repo_id, base_dataset_path)
+        self._upload_config_files(repo_id, base_dataset_path)
+
+        url = f"https://huggingface.co/datasets/{repo_id}"
```
Comment (johnnygreco) [inline packages/data-designer/src/data_designer/integrations/huggingface/client.py]: super nit: should we pull out the prefix as a constant?

File: packages/data-designer/src/data_designer/integrations/huggingface/client.py [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+        """Upload processor output files.
+
+        Args:
+            repo_id: Hugging Face dataset repo ID
+            base_dataset_path: Path to dataset directory
+
+        Raises:
+            HuggingFaceUploadError: If upload fails
+        """
+        processors_folder = base_dataset_path / "processors-files"
```
Comment (johnnygreco) [inline packages/data-designer/src/data_designer/integrations/huggingface/client.py]: ideally, we'd get this path off the `ArtifactStorage` instance. Not sure if that is easily accessible from here

File: packages/data-designer/src/data_designer/integrations/huggingface/client.py [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+            HuggingFaceUploadError: If validation fails or upload encounters errors
+        """
+        logger.info(f"🤗 Uploading dataset to Hugging Face Hub: {repo_id}")
+
+        self._validate_repo_id(repo_id)
+        self._validate_dataset_path(base_dataset_path)
+
+        self._create_or_get_repo(repo_id, private=private)
+
+        logger.info(f"|-- {RandomEmoji.data()} Uploading dataset card...")
```
Comment (johnnygreco) [inline packages/data-designer/src/data_designer/integrations/huggingface/client.py]: nit: I think all these need a couple spaces
```suggestion
        logger.info(f"  |-- {RandomEmoji.data()} Uploading dataset card...")
```

File: packages/data-designer/src/data_designer/integrations/huggingface/dataset_card_template.md [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+  data_files: "data/*.parquet"
+  default: true
+{% if has_processors %}{% for processor_name in processor_names %}- config_name: {{ processor_name }}
+  data_files: "{{ processor_name }}/*.parquet"
+{% endfor %}{% endif %}
+---
+
+<div style="display: flex; justify-content: space-between; align-items: flex-end; width: 100%; margin-bottom: 1rem;">
+<h1 style="flex: 1; margin: 0;">{{ repo_id.split('/')[-1] | title }}</h1>
+<sub style="white-space: nowrap;">Made with ❤️ using 🎨 <a href="https://github.com/NVIDIA-NeMo/DataDesigner">NeMo Data Designer</a></sub>
```
Comment (johnnygreco) [inline packages/data-designer/src/data_designer/integrations/huggingface/dataset_card_template.md:18]: LOVE IT

File: packages/data-designer/pyproject.toml [hunk 8]
```diff
@@ -22,6 +22,7 @@ classifiers = [
 dependencies = [
     "data-designer-config",
     "data-designer-engine",
+    "huggingface-hub>=1.0.1,<2",
```
Comment (johnnygreco) [inline packages/data-designer/pyproject.toml]: isn't this already an engine dep?

Comment (nabinchha) [reply packages/data-designer/pyproject.toml]: oh right! removed in de61805b4f9c2a6b2f216c6395562e083a290e55

File: packages/data-designer/src/data_designer/integrations/huggingface/client.py (hunk 6 above)
Comment (nabinchha) [reply packages/data-designer/src/data_designer/integrations/huggingface/client.py]: thanks for the catch! Updated in f0e3fcb5f6cbe4bad73d6da465c25b08d317240d

File: packages/data-designer/src/data_designer/integrations/huggingface/client.py [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+        except Exception as e:
+            raise HuggingFaceUploadError(f"Failed to generate dataset card: {e}") from e
+
+        try:
+            card.push_to_hub(repo_id, repo_type="dataset")
+        except Exception as e:
+            raise HuggingFaceUploadError(f"Failed to push dataset card to hub: {e}") from e
+
+    @staticmethod
+    def _validate_repo_id(repo_id: str) -> None:
```
Comment (andreatnvidia) [inline packages/data-designer/src/data_designer/integrations/huggingface/client.py:314]: Maybe use `from huggingface_hub.utils._validators import validate_repo_id`? It is private, but apparently stable (it's typically applied by a decorator it seems).
It has some additional checks compared to this function.

File: packages/data-designer/src/data_designer/integrations/huggingface/client.py [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+
+        self._create_or_get_repo(repo_id, private=private)
+
+        logger.info(f"  |-- {RandomEmoji.data()} Uploading dataset card...")
+        try:
+            self._upload_dataset_card(repo_id, base_dataset_path, description)
+        except Exception as e:
+            raise HuggingFaceUploadError(f"Failed to upload dataset card: {e}") from e
+
+        self._upload_main_dataset_files(repo_id, base_dataset_path)
```
Comment (andreatnvidia) [inline packages/data-designer/src/data_designer/integrations/huggingface/client.py]: Should these other steps be inside a try/except as well, and if something goes wrong we instruct users to look at URL and check partial upload?

File: packages/data-designer/src/data_designer/integrations/huggingface/client.py (hunk 5 above)
Comment (nabinchha) [reply packages/data-designer/src/data_designer/integrations/huggingface/client.py]: Ah good catch. I pulled out vars in `ArtifactStorage` in 99c61fedafb4289d89aa57288d90b25aa44d627c to try to DRY out some stuff in here. Directly injecting use of ArtifactStorage object in here looks a bit messy so I disengaged.

File: packages/data-designer/src/data_designer/integrations/huggingface/client.py (hunk 4 above)
Comment (nabinchha) [reply packages/data-designer/src/data_designer/integrations/huggingface/client.py]: Yes, done in 3270332e9554358a2ea92ce833ed956f44d4971b

File: packages/data-designer/src/data_designer/integrations/huggingface/__init__.py (hunk 3 above)
Comment (nabinchha) [reply packages/data-designer/src/data_designer/integrations/huggingface/__init__.py]: Do you mean rename `HuggingFaceUploadError` to `HuggingFaceHubClientUploadError` here?

File: packages/data-designer-engine/src/data_designer/engine/dataset_builders/artifact_storage.py (hunk 2 above)
Comment (nabinchha) [reply packages/data-designer-engine/src/data_designer/engine/dataset_builders/artifact_storage.py:265]: `sdg.json` and `metadata.json` had different spacing (2/4), so just trying to make things consistent.

File: packages/data-designer/src/data_designer/integrations/huggingface/client.py (hunk 9 above)
Comment (nabinchha) [reply packages/data-designer/src/data_designer/integrations/huggingface/client.py:314]: We could but it is in a module that's supposed to be private. `_validators`

File: packages/data-designer/src/data_designer/integrations/huggingface/client.py (hunk 10 above)
Comment (nabinchha) [reply packages/data-designer/src/data_designer/integrations/huggingface/client.py]: Ah each of those method implementations are in their own try catch...

File: packages/data-designer/src/data_designer/integrations/huggingface/client.py (hunk 9 above)
Comment (davanstrien) [reply packages/data-designer/src/data_designer/integrations/huggingface/client.py:314]: cc @Wauplin, is this reasonably safe to use even though it's private?

Comment (Wauplin) [reply packages/data-designer/src/data_designer/integrations/huggingface/client.py:314]: I'd rather prefer if you could use `from huggingface_hub.utils import validate_hf_hub_args` instead. It does some additional checks (warns on `proxies`, `resume_download`, `force_filename`, `local_dir_use_symlinks` which is irrelevant for you) but these checks are very light and doesn't infer your logic. And the `repo_id` will be correctly validated using HF rules.

Comment (nabinchha) [reply packages/data-designer/src/data_designer/integrations/huggingface/client.py:314]: @Wauplin @davanstrien thanks for the suggestion! Looks like `huggingface_hub.utils.validate_repo_id` is [available](https://huggingface.co/docs/huggingface_hub/main/en/package_reference/utilities#huggingface_hub.utils.validate_repo_id) as well. Is this reliable? It's more light-weight without being a decorator.

File: packages/data-designer/src/data_designer/integrations/huggingface/__init__.py (hunk 3 above)
Comment (nabinchha) [reply packages/data-designer/src/data_designer/integrations/huggingface/__init__.py]: Made the change above in ead52f5d871dbf0b1354bb9dd8b5e3355b1c9c9e

File: packages/data-designer/src/data_designer/integrations/huggingface/client.py (hunk 9 above)
Comment (nabinchha) [reply packages/data-designer/src/data_designer/integrations/huggingface/client.py:314]: I updated this method to use `huggingface_hub.utils.validate_repo_id` with some additional enforcement in bc90dcb6895b925968b646b99063637f7efbff12. @andreatgretel this looks like the same one you mentioned but exposed publicly. Thanks for the suggestion!

File: packages/data-designer/src/data_designer/integrations/huggingface/client.py [hunk 11]
```diff
@@ -0,0 +1,419 @@
...
+        """
+        # Check if repo_id is empty
+        if not repo_id or not repo_id.strip():
+            raise HuggingFaceHubClientUploadError("repo_id must be a non-empty string")
+
+        # Check for exactly one slash (username/dataset-name format). This is not enforced by huggingface_hub's validator.
+        if repo_id.count("/") != 1:
+            raise HuggingFaceHubClientUploadError(
+                f"Invalid repo_id format: '{repo_id}'. Expected format: 'username/dataset-name'"
+            )
```
[the remark this answers is not in the document]
Comment (nabinchha) [reply packages/data-designer/src/data_designer/integrations/huggingface/client.py:331]: Not true.

File: packages/data-designer/src/data_designer/integrations/huggingface/client.py (hunk 9 above)
Comment (Wauplin) [reply packages/data-designer/src/data_designer/integrations/huggingface/client.py:314]: > Looks like huggingface_hub.utils.validate_repo_id is [available](https://huggingface.co/docs/huggingface_hub/main/en/package_reference/utilities#huggingface_hub.utils.validate_repo_id) as well. Is this reliable? It's more light-weight without being a decorator.

Good catch! Sorry forgot about this one but yes, this is the correct way of doing it! It's a public and documented method so do not worry getting it removed without notice.

File: packages/data-designer/src/data_designer/interface/results.py [hunk 12]
```diff
@@ -96,3 +97,46 @@ def get_path_to_processor_artifacts(self, processor_name: str) -> Path:
...
             raise ArtifactStorageError(f"Processor {processor_name} has no artifacts.")
         return self.artifact_storage.processors_outputs_path / processor_name
+
+    def push_to_hub(
+        self,
+        repo_id: str,
+        description: str,
+        *,
+        token: str | None = None,
+        private: bool = False,
```
Comment (davanstrien) [inline packages/data-designer/src/data_designer/interface/results.py:107]: Not that important, but it could also be nice to allow users to pass custom tags here too?

File: packages/data-designer/src/data_designer/interface/results.py [hunk 13]
```diff
@@ -96,3 +97,46 @@ def get_path_to_processor_artifacts(self, processor_name: str) -> Path:
...
+        - Metadata (metadata.json)
+        - Auto-generated dataset card (README.md)
+
+        Args:
+            repo_id: HuggingFace repo ID (e.g., "username/my-dataset")
+            description: Custom description text for the dataset card.
+                Appears after the title.
+            token: HuggingFace API token. If None, the token is automatically
+                resolved from HF_TOKEN environment variable or cached credentials
+                from `huggingface-cli login`.
```
Comment (davanstrien) [inline packages/data-designer/src/data_designer/interface/results.py]:
```suggestion
                from `hf auth login`.
```

Comment (davanstrien) [reply packages/data-designer/src/data_designer/interface/results.py]: https://huggingface.co/docs/huggingface_hub/guides/cli#hf-auth-login new CLI entrypoints

File: packages/data-designer/src/data_designer/integrations/huggingface/client.py [hunk 14]
```diff
@@ -0,0 +1,419 @@
...
+class HuggingFaceHubClient:
+    """Client for interacting with Hugging Face Hub to upload datasets."""
+
+    def __init__(self, token: str | None = None):
+        """Initialize Hugging Face Hub client.
+
+        Args:
+            token: Hugging Face API token. If None, the token is automatically
+                resolved from HF_TOKEN environment variable or cached credentials
+                from `huggingface-cli login`.
```
Comment (davanstrien) [inline packages/data-designer/src/data_designer/integrations/huggingface/client.py]:
```suggestion
                from `hf auth login`.
```

File: packages/data-designer/src/data_designer/integrations/huggingface/client.py [hunk 15]
```diff
@@ -0,0 +1,419 @@
...
+                repo_type="dataset",
+                exist_ok=True,
+                private=private,
+            )
+        except HfHubHTTPError as e:
+            if e.response.status_code == 401:
+                raise HuggingFaceHubClientUploadError(
+                    "Authentication failed. Please provide a valid Hugging Face token. "
+                    "You can set it via the token parameter or HF_TOKEN environment variable, "
+                    "or run 'huggingface-cli login'."
```
Comment (davanstrien) [inline packages/data-designer/src/data_designer/integrations/huggingface/client.py]:
```suggestion
                    "or run 'hf auth login'."
```

File: packages/data-designer-engine/src/data_designer/engine/dataset_builders/artifact_storage.py [hunk 16]
```diff
@@ -25,6 +25,9 @@
 
 BATCH_FILE_NAME_FORMAT = "batch_{batch_number:05d}.parquet"
 SDG_CONFIG_FILENAME = "sdg.json"
```
Comment (johnnygreco) [inline packages/data-designer-engine/src/data_designer/engine/dataset_builders/artifact_storage.py]: nit: it still bothers me that it isn't clear what this file is... like is it  `config_builder.json`? I know BI uses that internally, but since we are making this config user-facing, it's a chance to be more explicit and consistent with DD?

File: packages/data-designer/src/data_designer/interface/results.py (hunk 13 above)
Comment (nabinchha) [reply packages/data-designer/src/data_designer/interface/results.py]: oh nice, thanks!

File: packages/data-designer/src/data_designer/interface/results.py (hunk 12 above)
Comment (nabinchha) [reply packages/data-designer/src/data_designer/interface/results.py:107]: that's a great idea! Added in e56c846c3af25a463a9d2ec04d8b88178b626b67

File: packages/data-designer-engine/src/data_designer/engine/dataset_builders/artifact_storage.py (hunk 16 above)
Comment (nabinchha) [reply packages/data-designer-engine/src/data_designer/engine/dataset_builders/artifact_storage.py]: sounds good. Updated to `builder_config.json` in 081ab2a6a413adc7304a2e766cc047261c034143

**Comment by johnnygreco** [comment]
Nice, thanks @nabinchha!

Thank you @davanstrien and @Wauplin for your feedback – super helpful!
